File: practice/algorithms/dynamic_programming/the_longest_increasing_subsequence/the_longest_increasing_subsequence.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> find_lis(const vector<int> & i_arr)
{
    if (i_arr.empty())
    {
        return vector<int>();
    }
    
    vector<int> p(i_arr.size());
    
    int u = 0, v = 0;
    
    vector<int> b(1);
    for (int pos = 1; pos < i_arr.size(); ++pos)
    {
        if (i_arr[b[b.size() - 1]] < i_arr[pos])
        {
            p[pos] = b[b.size() - 1];
            b.push_back(pos);
            continue;
        }
        
        u = 0;
        v = b.size();
        while (u < v)
        {
            int c = (u + v) / 2;
            if (i_arr[b[c]] < i_arr[pos])
            {
                u = c + 1;
            }
            else
            {
                v = c;
            }
        }
        
        if (i_arr[pos] < i_arr[b[u]])
        {
            if (u > 0)
            {
                p[pos] = b[u - 1];
            }
            b[u] = pos;
        }
    }
    
    u = b.size() - 1;
    v = b[u];
    while (u > 0)
    {
        b[u] = v;
        v = p[v];
        u--;
    }
    
    return b;
}

int main()
{
    int n;
    cin >> n;
    
    vector<int> arr(n);
    for (int pos = 0; pos < n; ++pos)
    {
        cin >> arr[pos];
    }
    
    vector<int> lis = find_lis(arr);
    cout << lis.size() << endl;
    
    return 0;
}

File: practice/algorithms/dynamic_programming/the_longest_increasing_subsequence/the_longest_increasing_subsequence.py
#!/usr/bin/env python

def find_lis(i_arr):
    if not i_arr:
        return []
    
    p = [0 for _ in range(len(i_arr))]
    u = 0
    v = 0

    b = [0]
    for POS in range(1, len(i_arr)):
        if i_arr[b[len(b) - 1]] < i_arr[POS]:
            p[POS] = b[len(b) - 1]
            b.append(POS)
            continue
        u = 0
        v = len(b)
        while u < v:
            c = (u + v) / 2
            if i_arr[b[c]] < i_arr[POS]:
                u = c + 1
            else:
                v = c
        if i_arr[POS] < i_arr[b[u]]:
            if u > 0:
                p[POS] = b[u - 1]
            b[u] = POS
    
    u = len(b) - 1
    v = b[-1]
    while u > 0:
        b[u] = v
        v = p[v]
        u -= 1
    return b
	
n = int(raw_input().strip())
arr = []
for _ in range(n):
    arr.append(int(raw_input().strip()))

print len(find_lis(arr))

